src/geometry/frustum.rs

```diff
@@ -1,67 +1,190 @@
 //! An asymmetric frustum with an arbitrary 3D pose.
 
-use crate::math::sat::{ConvexPolyhedron, Intersector};
-use crate::math::{Isometry3, PointCulling};
+use crate::geometry::Aabb;
+use crate::math::sat::{CachedAxesIntersector, ConvexPolyhedron, Intersector, Relation};
+use crate::math::PointCulling;
 use arrayvec::ArrayVec;
-use cgmath::{
-    BaseFloat, Decomposed, InnerSpace, Matrix4, Perspective, Point3, Quaternion, Transform, Vector3,
-};
-use collision::{Aabb3, Relation};
+use nalgebra::{Isometry3, Matrix4, Point3, RealField, Unit, Vector3};
+use num_traits::Bounded;
 use serde::{Deserialize, Serialize};
 
+/// A perspective projection matrix analogous to cgmath::Perspective.
+#[derive(Debug, Clone, Serialize, Deserialize)]
+pub struct Perspective<S: RealField> {
+    matrix: Matrix4<S>,
+}
+
+impl<S: RealField> Perspective<S> {
+    /// Left, right, bottom, and top are in radians.
+    pub fn new(left: S, right: S, bottom: S, top: S, near: S, far: S) -> Self {
+        assert!(
+            left <= right,
+            "`left` cannot be greater than `right`, found: left: {:?} right: {:?}",
+            left,
+            right
+        );
+        assert!(
+            bottom <= top,
+            "`bottom` cannot be greater than `top`, found: bottom: {:?} top: {:?}",
+            bottom,
+            top
+        );
+        assert!(
+            near <= far,
+            "`near` cannot be greater than `far`, found: near: {:?} far: {:?}",
+            near,
+            far
+        );
+
+        let two: S = nalgebra::convert(2.0);
+
+        let r0c0 = (two * near) / (right - left);
+        let r0c1 = nalgebra::zero();
+        let r0c2 = (right + left) / (right - left);
+        let r0c3 = nalgebra::zero();
+
+        let r1c0 = nalgebra::zero();
+        let r1c1 = (two * near) / (top - bottom);
+        let r1c2 = (top + bottom) / (top - bottom);
+        let r1c3 = nalgebra::zero();
+
+        let r2c0 = nalgebra::zero();
+        let r2c1 = nalgebra::zero();
+        let r2c2 = -(far + near) / (far - near);
+        let r2c3 = -(two * far * near) / (far - near);
+
+        let r3c0 = nalgebra::zero();
+        let r3c1 = nalgebra::zero();
+        let r3c2 = -S::one();
+        let r3c3 = nalgebra::zero();
+
+        #[rustfmt::skip]
+        let matrix = Matrix4::new(
+            r0c0, r0c1, r0c2, r0c3,
+            r1c0, r1c1, r1c2, r1c3,
+            r2c0, r2c1, r2c2, r2c3,
+            r3c0, r3c1, r3c2, r3c3,
+        );
+        Self { matrix }
+    }
+
+    // This emulates cgmath::PerspectiveFov, which is more restricted
+    // and corresponds to nalgebra::Perspective.
+    pub fn new_fov(fovy: S, aspect: S, near: S, far: S) -> Self {
```

Comment:
nit: since we are using `nalgebra` now, should we reorder the parameters to be on par with https://www.nalgebra.org/rustdoc/nalgebra/geometry/struct.Perspective3.html#method.new ?
i.e.
```suggestion
    pub fn new_fov(aspect: S, fovy: S, near: S, far: S) -> Self {
```

Reply:
Hmm, no strong opinions. Having this be a drop-in for cgmath::Perspective makes the diff smaller. And I'd have to be careful when reordering those as they are of the same type.

Reply:
For my own understanding and yours, let's take a step back and look at where PerspectiveFOV was used to understand how to replace it:
* In `point_cloud_test` where it's immediately converted to the more general `Perspective`
* In `point_viewer_grpc` where it's immediately converted to the more general `Perspective`
* In `sdl_viewer` where it's immediately converted to a matrix
* (avmapping-geometry) In `point_cloud_client_cc` where it's immediately converted to a matrix

We can replace the latter two with `nalgebra::Perspective3`, but for the former two we need conversion from the special case (symmetric frustum)  to the general case (asymmetric frustum, i.e. `Perspective`). I did that via a "convenience constructor" but thinking about it again, I could implement a conversion from `nalgebra::Perspective3`. We don't need to define a new struct whose only purpose is to convert to `Perspective` at any rate.

Reply:
Done, I added a `From<nalgebra::Perspective3>` and added a test that checks that it's equal to `new_fov` :)